README.md
# kgp-150-leak-patcher

A simple plugin to automatically patch the memory leak in Kotlin Gradle Plugin 1.5.0 in 
https://youtrack.jetbrains.com/issue/KT-46368.

Implementation based on [David Burström](https://github.com/davidburstrom)'s workaround [here](https://youtrack.jetbrains.com/issue/KT-46368#focus=Comments-27-4868598.0-0).

## Usage

Apply the plugin in the root `build.gradle` file, no other configuration needed. It is only 
applicable for KGP 1.5.0 (not 1.4.32, should be fixed in 1.5.10). Any other version will print
a warning and no-op.

The minimum required Gradle version is `6.1`.

```kotlin
plugins {
  id("dev.zacsweers.kgp-150-leak-patcher") version "1.1.0"
}

// Or legacy way
buildscript {
  dependencies {
    classpath("dev.zacsweers:kgp-150-leak-patcher:1.1.0")
  }
}

apply(plugin = "dev.zacsweers.kgp-150-leak-patcher")
```

If you need to disable the applicability check for some reason, it can be disabled by setting the
`kgp15leak.checkApplicability` property to `false`.

Snapshots of the development version are available in [Sonatype's `snapshots` repository][snapshots].

License
-------

    Copyright (C) 2021 Zac Sweers

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.

[snapshots]: https://oss.sonatype.org/content/repositories/snapshots/


CHANGELOG.md
Changelog
=========

1.1.0
-----

_2021-05-08_

**New:** The plugin now uses a Gradle `BuildService` under the hood, which makes it 
compatible with Configuration Caching. Thanks to 
[@christiandeange](https://github.com/christiandeange) from Square for contributing this!

**New:** If you need to disable the applicability check for some reason, it can be disabled by 
setting the `kgp15leak.checkApplicability` property to `false`.

1.0.1
-----

_2021-05-05_

**Fix:** Inverted logic on the plugin application in a recent refactoring meant the plugin wasn't getting applied.

Thanks to [@msfjarvis](https://github.com/msfjarvis) for this quick fix!

1.0.0
-----

_2021-05-05_

Initial release.


src/test/kotlin/dev/zacsweers/kgp/EmbeddableVersionCheckerTest.kt
package dev.zacsweers.kgp

import com.google.common.truth.Truth.assertThat
import org.gradle.util.VersionNumber
import org.junit.Test

class EmbeddableVersionCheckerTest {

  @Test
  fun simpleTest() {
    val version = parseCompilerEmbeddedVersionNumber(
      "1.5.0-release-749 (1.5.0)"
    )
    assertThat(version).isEqualTo(VersionNumber.parse("1.5.0"))
    assertThat(isApplicable(version)).isTrue()

    val newVersion = parseCompilerEmbeddedVersionNumber(
      "1.5.10-RC-release-749 (1.5.10)"
    )
    assertThat(newVersion).isEqualTo(VersionNumber.parse("1.5.10"))
    assertThat(isApplicable(newVersion)).isFalse()

    val oldVersion = parseCompilerEmbeddedVersionNumber(
      "1.4.32-RC-release-749 (1.4.32)"
    )
    assertThat(oldVersion).isEqualTo(VersionNumber.parse("1.4.32"))
    assertThat(isApplicable(oldVersion)).isFalse()
  }

  @Test
  fun dynamicTest() {
    // Compute based on CI test
    val target = System.getenv().getOrDefault("ci_kotlin_version", "1.5.0")
    val version = parseCompilerEmbeddedVersionNumber()
    assertThat(version).isEqualTo(VersionNumber.parse(target))
  }

  @Test
  fun applicability() {
    assertThat(shouldCheckApplicability("true")).isTrue()
    assertThat(shouldCheckApplicability("false")).isFalse()
    assertThat(shouldCheckApplicability(null)).isTrue()
  }
}

src/main/kotlin/dev/zacsweers/kgp/Kgp150LeakPatcherBuildService.kt
package dev.zacsweers.kgp

import org.gradle.api.services.BuildService
import org.gradle.api.services.BuildServiceParameters
import org.jetbrains.kotlin.cli.common.CompilerSystemProperties

/**
 * By implementing [AutoCloseable], Gradle will call the [close] method at the end of the build.
 */
internal abstract class Kgp150LeakPatcherBuildService : BuildService<BuildServiceParameters.None>, AutoCloseable {

  override fun close() {
    CompilerSystemProperties.systemPropertyGetter = SystemPropertyGetter()
    CompilerSystemProperties.systemPropertySetter = SystemPropertySetter()
    CompilerSystemProperties.systemPropertyCleaner = SystemPropertyCleaner()
  }
}

private class SystemPropertyGetter : (String) -> String? {
  override operator fun invoke(p1: String): String? = System.getProperty(p1)
}

private class SystemPropertySetter : (String, String) -> String? {
  override operator fun invoke(o1: String, o2: String): String = System.setProperty(o1, o2)
}

private class SystemPropertyCleaner : (String) -> String? {
  override operator fun invoke(p1: String): String = System.clearProperty(p1)
}


src/main/kotlin/dev/zacsweers/kgp/Kgp150LeakPatcherPlugin.kt
package dev.zacsweers.kgp

import org.gradle.api.Plugin
import org.gradle.api.Project
import org.gradle.util.VersionNumber
import org.jetbrains.kotlin.cli.common.CompilerSystemProperties

/**
 * Cover for a memory leak in KGP 1.5.0 - https://youtrack.jetbrains.com/issue/KT-46368
 *
 * Remove in 1.5.10!
 */
class Kgp150LeakPatcherPlugin : Plugin<Project> {
  override fun apply(project: Project) {
    if (shouldCheckApplicability(project.findProperty(APPLICABILITY_CHECK_PROP))) {
      val embeddableVersion = parseCompilerEmbeddedVersionNumber()
      if (!isApplicable(embeddableVersion)) {
        project.logger.warn("KGP 1.5.0 Leak Patcher plugin is only applicable to Kotlin 1.5.0. Detected version '$embeddableVersion'.")
        return
      }
    }

    project.gradle.sharedServices.registerIfAbsent("kgp-150-leak-patcher", Kgp150LeakPatcherBuildService::class.java) {}
      // Force the BuildService to be loaded into memory immediately.
      .get()
  }

  companion object {
    /** A boolean gradle property to disable applicability checks. Use at your own risk! */
    const val APPLICABILITY_CHECK_PROP = "kgp15leak.checkApplicability"
  }
}

private val KGP_150 = VersionNumber.parse("1.5.0")

internal fun shouldCheckApplicability(propValue: Any?): Boolean {
  return propValue?.toString()
    ?.toBoolean()
    ?: true
}

internal fun isApplicable(versionNumber: VersionNumber = parseCompilerEmbeddedVersionNumber()): Boolean {
  return versionNumber == KGP_150
}

private fun parseCompilerEmbeddedVersion(): String? {
  // implementationVersion is like '1.5.0-release-749 (1.5.0)'
  return try {
    CompilerSystemProperties::class.java.`package`.implementationVersion
  } catch (e: NoClassDefFoundError) {
    // Cover for old plugins
    return null
  }
}

internal fun parseCompilerEmbeddedVersionNumber(
  version: String? = parseCompilerEmbeddedVersion()
): VersionNumber {
  return version?.substringBefore("-")
    ?.let(VersionNumber::parse)
    ?.baseVersion
    ?: VersionNumber.UNKNOWN
}
